rts/Lua/LuaSyncedRead.cpp

```diff
@@ -8515,8 +8520,139 @@ int LuaSyncedRead::GetUnitScriptNames(lua_State* L)
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	QuadFieldQuery qfQuery;
+	quadField.GetQuadsOnRay(qfQuery, pos, dir, traceLength);
+	CollisionQuery cq;
+	int num = 0;
+	lua_newtable(L);
+	for (const int quadIdx : *qfQuery.quads) {
+		const CQuadField::Quad& quad = quadField.GetQuad(quadIdx);
 
+		for (CUnit* u : quad.units) {
 
+			if (!u->HasCollidableStateBit(CSolidObject::CSTATE_BIT_QUADMAPRAYS))
+				continue;
+
+			if (CCollisionHandler::DetectHit(u, u->GetTransformMatrix(true), pos, pos + dir * traceLength, &cq, true)) {
+				const float len = cq.GetHitPosDist(pos, dir);
+				lua_newtable(L);
+				lua_pushnumber(L, len);
+				lua_rawseti(L, -2, 1);
+				lua_pushnumber(L, u->id);
+				lua_rawseti(L, -2, 2);
+				lua_rawseti(L, -2, num + 1);
+				num++;
+			}
+		}
+	}
+	return 1;
+}
+/***
+ *
+ * @function Spring.TraceRayFeatures
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @treturn {{len, Id},...}
+ */
+int LuaSyncedRead::TraceRayFeatures(lua_State* L) //returns the list of features that an raytrace has hit
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	QuadFieldQuery qfQuery;
+	quadField.GetQuadsOnRay(qfQuery, pos, dir, traceLength);
+	CollisionQuery cq;
+	int num = 0;
+	lua_newtable(L);
+	for (const int quadIdx : *qfQuery.quads) {
+		const CQuadField::Quad& quad = quadField.GetQuad(quadIdx);
+
+		for (CFeature* f : quad.features) {
+			if (!f->HasCollidableStateBit(CSolidObject::CSTATE_BIT_QUADMAPRAYS))
+				continue;
+
+			if (CCollisionHandler::DetectHit(f, f->GetTransformMatrix(true), pos, pos + dir * traceLength, &cq, true)) {
+				const float len = cq.GetHitPosDist(pos, dir);
+				lua_newtable(L);
+				lua_pushnumber(L, len);
+				lua_rawseti(L, -2, 1);
+				lua_pushnumber(L, f->id);
+				lua_rawseti(L, -2, 2);
+				lua_rawseti(L, -2, num + 1);
+				num++;
+			}
+		}
+	}
+	return 1;
+}
+/***
+ *
+ * @function Spring.TraceRayGround
+ * @number posX
+ * @number posY
+ * @number posZ
+ * @number dirX
+ * @number dirY
+ * @number dirZ
+ * @number traceLength
+ * @bool[opt=false] ignoreWater
+ * @treturn nil|len
+ * @treturn number     posX
+ * @treturn number     posY
+ * @treturn number     posZ
+ */
+int LuaSyncedRead::TraceRayGround(lua_State* L)
+{
+	float3 pos(luaL_checkfloat(L, 1), luaL_checkfloat(L, 2), luaL_checkfloat(L, 3));
+	float3 dir(luaL_checkfloat(L, 4), luaL_checkfloat(L, 5), luaL_checkfloat(L, 6));
+	float traceLength = luaL_checkfloat(L, 7);
+	float3 normDir = dir.Normalize();
+	auto result2 = pos + normDir * traceLength;
+	const float groundLength = CGround::LineGroundCol(pos, result2);
```

Comment:
This part should probably keep using un-normalized dir so that you can do the (end - start) trick. Also maybe it starts obeying LoS if you just pass the sync value of the handle as the 3rd?
```suggestion
	const float groundLength = CGround::LineGroundCol(pos, pos + dir * traceLength, CLuaHandle::GetHandleSynced(L));
```